File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/views/show_tea/countdowntimer/BackgroundTimerTest.java
package coolpharaoh.tee.speicher.tea.timer.views.show_tea.countdowntimer;

import android.app.AlarmManager;
import android.app.Application;
import android.app.PendingIntent;
import android.content.Context;
import android.os.Build;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

//could be removed when Robolectric supports Java 8 for API 29
@Config(sdk = Build.VERSION_CODES.O_MR1)
@RunWith(RobolectricTestRunner.class)
public class BackgroundTimerTest {
    @Rule
    public MockitoRule rule = MockitoJUnit.rule();
    @Mock
    Application application;
    @Mock
    SharedTimerPreferences sharedTimerPreferences;
    @Mock
    AlarmManager alarmManager;

    @Test
    public void setAlarmManager() {
        when(sharedTimerPreferences.getStartedTime()).thenReturn(2000L);
        when(application.getSystemService(Context.ALARM_SERVICE)).thenReturn(alarmManager);

        BackgroundTimer backgroundTimer = new BackgroundTimer(application, sharedTimerPreferences);
        backgroundTimer.setAlarmManager(1L, 6000L);

        ArgumentCaptor<AlarmManager.AlarmClockInfo> captorAlarmClockInfo = ArgumentCaptor.forClass(AlarmManager.AlarmClockInfo.class);
        verify(alarmManager).setAlarmClock(captorAlarmClockInfo.capture(), any(PendingIntent.class));

        AlarmManager.AlarmClockInfo alarmClockInfo = captorAlarmClockInfo.getValue();
        assertThat(alarmClockInfo.getTriggerTime()).isEqualTo(8000L);
        assertThat(alarmClockInfo.getShowIntent()).isNotNull();
    }

    @Test(expected = NullPointerException.class)
    public void whenSetAlarmManagerAndItIsNullThrowException() {
        when(application.getSystemService(Context.ALARM_SERVICE)).thenReturn(null);

        BackgroundTimer backgroundTimer = new BackgroundTimer(application, sharedTimerPreferences);
        backgroundTimer.setAlarmManager(1L, 6000L);

        verify(alarmManager, times(0)).setAlarmClock(any(), any());
    }

    @Test
    public void removeAlarmManager() {
        when(application.getSystemService(Context.ALARM_SERVICE)).thenReturn(alarmManager);

        BackgroundTimer backgroundTimer = new BackgroundTimer(application, sharedTimerPreferences);
        backgroundTimer.removeAlarmManager();

        verify(alarmManager).cancel(any(PendingIntent.class));
    }

    @Test(expected = NullPointerException.class)
    public void whenRemoveAlarmManagerAndItIsNullThrowException() {
        when(application.getSystemService(Context.ALARM_SERVICE)).thenReturn(null);

        BackgroundTimer backgroundTimer = new BackgroundTimer(application, sharedTimerPreferences);
        backgroundTimer.removeAlarmManager();

        verify(alarmManager, times(0)).setAlarmClock(any(), any());
    }
}

File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/views/show_tea/countdowntimer/BackgroundTimerTest.kt
package coolpharaoh.tee.speicher.tea.timer.views.show_tea.countdowntimer

import android.app.AlarmManager
import android.app.AlarmManager.AlarmClockInfo
import android.app.Application
import android.app.PendingIntent
import android.content.Context
import io.mockk.every
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit4.MockKRule
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions.*
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class BackgroundTimerTest {
    @get:Rule
    val mockkRule = MockKRule(this)
    @RelaxedMockK
    lateinit var application: Application
    @RelaxedMockK
    lateinit var sharedTimerPreferences: SharedTimerPreferences
    @RelaxedMockK
    lateinit var alarmManager: AlarmManager

    @Test
    fun setAlarmManager() {
        every { sharedTimerPreferences.startedTime } returns 2000L
        every { application.getSystemService(Context.ALARM_SERVICE) } returns alarmManager

        val backgroundTimer = BackgroundTimer(application, sharedTimerPreferences)
        backgroundTimer.setAlarmManager(1L, 6000L)

        val slotAlarmClockInfo = slot<AlarmClockInfo>()
        verify { alarmManager.setAlarmClock(capture(slotAlarmClockInfo),any()) }

        val alarmClockInfo = slotAlarmClockInfo.captured
        assertThat(alarmClockInfo.triggerTime).isEqualTo(8000L)
        assertThat(alarmClockInfo.showIntent).isNotNull
    }

    @Test(expected = NullPointerException::class)
    fun whenSetAlarmManagerAndItIsNullThrowException() {
        every { application.getSystemService(Context.ALARM_SERVICE) } returns null

        val backgroundTimer = BackgroundTimer(application, sharedTimerPreferences)
        backgroundTimer.setAlarmManager(1L, 6000L)

        verify (exactly = 0) { alarmManager.setAlarmClock(any(), any()) }
    }

    @Test
    fun removeAlarmManager() {
        every { application.getSystemService(Context.ALARM_SERVICE) } returns alarmManager

        val backgroundTimer = BackgroundTimer(application, sharedTimerPreferences)
        backgroundTimer.removeAlarmManager()

        verify { alarmManager.cancel(any<PendingIntent>()) }
    }

    @Test(expected = NullPointerException::class)
    fun whenRemoveAlarmManagerAndItIsNullThrowException() {
        every { application.getSystemService(Context.ALARM_SERVICE) } returns null

        val backgroundTimer = BackgroundTimer(application, sharedTimerPreferences)
        backgroundTimer.removeAlarmManager()

        verify (exactly = 1) { alarmManager.setAlarmClock(any(), any()) }
    }
}